pkg/controller/framework/shared_informer.go

```diff
@@ -0,0 +1,243 @@
+	GetController() ControllerInterface
+	Run(stopCh <-chan struct{})
+	HasSynced() bool
+}
+
+type SharedIndexInformer interface {
+	SharedInformer
+
+	AddIndexer(indexer cache.Indexer) error
+	GetIndexer() cache.Indexer
+}
+
+func NewSharedInformer(lw cache.ListerWatcher, objType runtime.Object, resyncPeriod time.Duration) SharedInformer {
+	sharedInformer := &ShareableInformer{
+		processor: &shareableProcessor{},
+		store:     cache.NewStore(DeletionHandlingMetaNamespaceKeyFunc),
+	}
+
+	// This will hold incoming changes. Note how we pass store in as a
+	// KeyLister, that way resync operations will result in the correct set
+	// of update/delete deltas.
+	fifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, sharedInformer.store)
+
+	cfg := &Config{
+		Queue:            fifo,
+		ListerWatcher:    lw,
+		ObjectType:       objType,
+		FullResyncPeriod: resyncPeriod,
+		RetryOnError:     false,
+
+		Process: sharedInformer.HandleDeltas,
+	}
+	sharedInformer.controller = New(cfg)
+
+	return sharedInformer
+}
+
+type ShareableInformer struct {
+	store      cache.Store
+	controller *Controller
+
+	processor *shareableProcessor
+
+	runVotes    sync.WaitGroup
+	started     bool
+	startedLock sync.Mutex
+}
+
+type votingController struct {
+	informer *ShareableInformer
+}
+
+func (v *votingController) Run(stopCh <-chan struct{}) {
+	v.informer.runVotes.Done()
+}
+
+func (v *votingController) HasSynced() bool {
+	return v.informer.HasSynced()
+}
+
+type shareableProcessor struct {
+	listeners []*processorListener
+}
+
+const (
+	channelDepth = 1000
+)
+
+type processorListener struct {
+	notifications chan interface{}
+
+	handler ResourceEventHandler
+}
+
+type updateNotification struct {
+	oldObj interface{}
+	newObj interface{}
+}
+
+type addNotification struct {
+	newObj interface{}
+}
+
+type deleteNotification struct {
+	oldObj interface{}
+}
+
+func NewProcessListener(handler ResourceEventHandler) *processorListener {
+	return &processorListener{
+		notifications: make(chan interface{}, channelDepth),
+		handler:       handler,
+	}
+}
+
+func (s *ShareableInformer) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	func() {
+		s.startedLock.Lock()
+		defer s.startedLock.Unlock()
+		s.started = true
+	}()
+
+	s.runVotes.Wait()
+
+	s.processor.Run(stopCh)
+	s.controller.Run(stopCh)
+}
+
+func (s *ShareableInformer) isStarted() bool {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+	return s.started
+}
+
+func (s *ShareableInformer) HasSynced() bool {
+	return s.controller.HasSynced()
+}
+
+func (s *ShareableInformer) GetStore() cache.Store {
+	return s.store
+}
+
+func (s *ShareableInformer) GetController() ControllerInterface {
+	return &votingController{informer: s}
+}
+
+func (s *ShareableInformer) AddEventHandler(handler ResourceEventHandler) error {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+
+	if s.started {
+		return fmt.Errorf("informer has already started")
```

Comment:
OK - so you are explicitly disallowing registering handler functions after starting an informer.
This seems ok for our usecases, but we need to explicitly document it.